### PR TITLE
Speed-up 5-arg mul! with scalar and equal-axes arrays

### PR DESCRIPTION
**Pull request by dkarrasch**

### CODE REVIEW

File: stdlib/LinearAlgebra/src/generic.jl [hunk 1]
```diff
@@ -121,10 +129,22 @@ match the length of the second, $(length(X))."))
     C
 end
 
-@inline mul!(C::AbstractArray, s::Number, X::AbstractArray, alpha::Number, beta::Number) =
-    generic_mul!(C, s, X, MulAddMul(alpha, beta))
-@inline mul!(C::AbstractArray, X::AbstractArray, s::Number, alpha::Number, beta::Number) =
-    generic_mul!(C, X, s, MulAddMul(alpha, beta))
+@inline function mul!(C::AbstractArray, s::Number, X::AbstractArray, alpha::Number, beta::Number)
+    if axes(C) == axes(X)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/generic.jl:133]: Is there a weaker test that would ensure that broadcasting works (something with `size`?), or is this just right? @mbauman

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/generic.jl:133]: This seems to be correct. You can't broadcast a regular array with an OffsetArray of same size, so indeed axes must match.

Comment (mcabbott) [reply stdlib/LinearAlgebra/src/generic.jl:133]: Some `mul!` methods call `require_one_based_indexing`. It's possible that this should too for consistency, not sure. 

Ultimately it would be nice to allow offsets, e.g.  https://github.com/JuliaArrays/OffsetArrays.jl/issues/91

Comment (mcabbott) [reply stdlib/LinearAlgebra/src/generic.jl:133]: Perhaps I am being dense but what's an example of input that `generic_mul!` accepts, with `s::Number`, which can't be done by broadcasting?

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/generic.jl:133]: I agree. The condition here is just to make sure the broadcasting in the line below works. And for genericity, I think that one should not require one-based indexing, right? In view of OffsetArrays, would you neglect the axes and allow inconsistent axes in scalar `mul!`?

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/generic.jl:133]: > Perhaps I am being dense but what's an example of input that `generic_mul!` accepts, with `s::Number`, which can't be done by broadcasting?

Well, if they have different size, to begin with. And second, I think broadcasting throws if you mix a regular array with an offset array, unless I have messed something up.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/generic.jl:133]:
```julia
julia> A = rand(2,3)
2×3 Matrix{Float64}:
 0.126249  0.593993  0.968903
 0.522808  0.785498  0.51491

julia> B = OffsetArray(rand(2,3), 0:1, 0:2)
2×3 OffsetArray(::Matrix{Float64}, 0:1, 0:2) with eltype Float64 with indices 0:1×0:2:
 0.290057  0.41911   0.501342
 0.724973  0.470591  0.00700215

julia> A .+ B
ERROR: DimensionMismatch("arrays could not be broadcast to a common size; got a dimension with lengths 2 and 2")
```

Comment (mcabbott) [reply stdlib/LinearAlgebra/src/generic.jl:133]: Right, broadcasting requires that `axes` agree. 

But `*` and `mul!` right now often require something stronger, that they agree and start at 1. It would be nice to relax this, to allow `*` between OffsetArrays, but this would still require agreement of axes between arrays. Until that happens more widely, should methods which can easily support it (like this one) do so, or is that inconsistency weird? I guess the answer is that they should, going by this (which is also broadcast not BLAS):
```
julia> [1,100] * OffsetArray(1:3,5)'
2×3 OffsetArray(::Array{Int64,2}, 1:2, 6:8) with eltype Int64 with indices 1:2×6:8:
   1    2    3
 100  200  300
```

My other question (and perhaps this is orthogonal) is whether the `else` clause is needed at all. It seems like whenever `mul!(C, s, X, α, β)` with scalar `s` makes sense, it can be written as a broadcast. If `C` is immutable, then neither path will work. Or can  `X` have a shape which `generic_mul!` allows & `.=` does not?

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/generic.jl:133]: Take a look at the `generic_mul!(C, s, X, MulAddMul(alpha, beta))` that used to be called (and still is in case the axes don't match): it just loops over the `zip` of the indices, and hence doesn't even require the arrays to be of equal dimension.
```julia
julia> C = zeros(10); A = ones(2,5); mul!(C, 2, A, true, true)
10-element Array{Float64,1}:
 2.0
...
```
But
```julia
julia> 2 .*A.*true .+ C.*true
ERROR: DimensionMismatch("arrays could not be broadcast to a common size; got a dimension with lengths 2 and 10")
```
This method is so generous, it works for higher-dimensional arrays than matrices, for arrays of different `ndims`, etc.

I realize that I should have posted an issue first, demonstrating what is bad and what is being fixed/improved here (from Slack):
```julia
n = 1024
x = ones(n)
v = zeros(n)
D = Diagonal(ones(n))
M = 1.0I
@benchmark mul!($v, $D, $x)
memory estimate:  0 bytes
allocs estimate:  0
--------------
minimum time:     104.696 ns (0.00% GC)
median time:      107.364 ns (0.00% GC)
mean time:        113.303 ns (0.00% GC)
maximum time:     441.836 ns (0.00% GC)
--------------
samples:          10000
evals/sample:     937

@benchmark mul!($v, $M, $x)
BenchmarkTools.Trial:
memory estimate:  0 bytes
allocs estimate:  0
--------------
minimum time:     382.759 ns (0.00% GC)
median time:      395.074 ns (0.00% GC)
mean time:        413.855 ns (0.00% GC)
maximum time:     1.140 μs (0.00% GC)
--------------
samples:          10000
evals/sample:     203
```
With the PR here, the `Number` case is now faster than the `Diagonal` one, as expected.

In summary, this PR doesn't restrict anything, it speeds things up in more specific cases.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/generic.jl:133]: Wait, and in fact, `generic_mul!` ("5-arg `mul!`") _can_ handle OffsetArrays. It's probably ridiculously slow, but still. With `A = OffsetArray(rand(2,5), 0:1, 0:4)` and `C` as before, you can do `mul!(C, 2, A, true, true)`.

Comment (mcabbott) [reply stdlib/LinearAlgebra/src/generic.jl:133]: Oh that's awful, why is `LinearAlgebra.mul!` treating these things as bags of numbers? Can we fix that, is it a breaking change or a bugfix? Maybe it has to be filed next to recursive `dot` as something to fix in 2.0... 

And in your example, `M::UniformScaling{Float64}` does come to this scalar method, OK. For sane input, it will hit the broadcast path.

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/generic.jl:133]: This method is the most generic fallback, and it's been there since ages, I guess. Before the 5-arg era, there was a corresponding 3-arg version `mul!(C, 3, A)` (without the `UniformScaling` stuff, apparently). I wasn't aware of that before I was looking into this `Number`-based generic multiplication code. So, seems like somebody put it there on purpose. I'm not sure either why or where one would do such "mathematical nonsense".
